src/hotspot/share/oops/method.cpp

```diff
@@ -2254,10 +2254,12 @@ bool Method::is_method_id(jmethodID mid) {
 Method* Method::checked_resolve_jmethod_id(jmethodID mid) {
   if (mid == NULL) return NULL;
   Method* o = resolve_jmethod_id(mid);
-  if (o == NULL || o == JNIMethodBlock::_free_method || !((Metadata*)o)->is_method()) {
+  if (o == NULL || o == JNIMethodBlock::_free_method) {
     return NULL;
   }
-  return o;
+  // Method should otherwise be valid. Assert for testing.
+  assert(is_valid_method(o), "should be valid jmethodid");
```

Comment:
nit typo: s/jmethodid/jmethodID/

Reply:
I think you need a comment for L2262 to explain why you return NULL
when the class loader is no longer alive. I think it's because it's racy
if the class loader is no longer alive to return the jmethodID.

However, if that is racy, then does that mean that the above new assert()
is also racy?

Reply:
The assert isn't racy because there needs to be a safepoint in between changing the color of the oop pointers, so if there's a Method there, the class loader liveness won't change state in between.  The assert checks for the unexpected case where it's corrupted.